File: leetcode/113. Path Sum II.py
# Given a binary tree and a sum, find all root-to-leaf paths where each path's sum equals the given sum.

# Note: A leaf is a node with no children.

# Example:

# Given the below binary tree and sum = 22,

#       5
#      / \
#     4   8
#    /   / \
#   11  13  4
#  /  \    / \
# 7    2  5   1
# Return:

# [
#    [5,4,11,2],
#    [5,8,4,5]
# ]
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int):
        res = []
        ans = []
        self.pathSumDFS(root,sum,ans,res)
        return res    
    def pathSumDFS(self,root,target,ans,res):
        if root == None:
            return 0
        ans.append(root.val)
        if root.val == target and root.left == None and root.right ==None:
            res.append(ans.copy())
        self.pathSumDFS(root.left,target-root.val,ans.copy(),res)
        self.pathSumDFS(root.right,target-root.val,ans.copy(),res)
        ans.pop()
 

File: leetcode/113. Path Sum II.cpp
#include <iostream>
#include <unordered_set>
#include <algorithm>
#include <vector>
#include <string>
#include <numeric>
#include <queue>
using namespace std;



// Definition for a binary tree node.
 struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

 class Solution {
 public:
	 vector<vector<int>> pathSum(TreeNode* root, int sum) {
		 vector<vector<int>> res;
		 vector<int> out;
		 help(root, sum, out, res);
		 return res;
	 }
	 void help(TreeNode* node, int sum, vector<int>& out, vector<vector<int>>res) {
		 if (!node) return;
		 out.push_back(node->val);
		 if (sum == node->val && !node->left && !node->right)
			 res.push_back(out);
		 help(node->left, sum - node->val, out, res);
		 help(node->right, sum - node->val, out, res);
		 out.pop_back();
	 }
 };
int main() {
	/*vector<int> i = {1, 2, 3, 4, 0};
	Solution s;
	vector<int> ans = s.singleNumber(i);
	cout<<ans[0]<<ans[1]<< endl;*/
	system("pause");
	return 0;
}